src/net/addr.rs

```diff
@@ -0,0 +1,159 @@
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::net::{SocketAddr, SocketAddrV4, SocketAddrV6};
+use std::pin::Pin;
+
+use cfg_if::cfg_if;
+use futures::future::{ready, Ready};
+
+use crate::future::Future;
+use crate::io;
+use crate::task::blocking;
+use crate::task::{Context, Poll};
+
+cfg_if! {
+    if #[cfg(feature = "docs")] {
+        #[doc(hidden)]
+        pub struct ImplFuture<T>(std::marker::PhantomData<T>);
+
+        macro_rules! ret {
+            ($f:tt, $i:ty) => (ImplFuture<io::Result<$i>>);
+        }
+    } else {
+        macro_rules! ret {
+            ($f:tt, $i:ty) => ($f<$i>);
+        }
+    }
+}
+
+/// A trait for objects which can be converted or resolved to one or more [`SocketAddr`] values.
+///
+/// This trait is an async version of [`std::net::ToSocketAddrs`].
+///
+/// [`std::net::ToSocketAddrs`]: https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html
+/// [`SocketAddr`]: https://doc.rust-lang.org/std/net/enum.SocketAddr.html
+pub trait ToSocketAddrs {
+    /// Returned iterator over socket addresses which this type may correspond to.
+    type Iter: Iterator<Item = SocketAddr> + Send;
+
+    /// Converts this object to an iterator of resolved `SocketAddr`s.
+    ///
+    /// The returned iterator may not actually yield any values depending on the outcome of any
+    /// resolution performed.
+    ///
+    /// Note that this function may block a backend thread while resolution is performed.
+    fn to_socket_addrs(&self) -> ret!(ToSocketAddrsFuture, Self::Iter);
```

Comment:
Let's parametrize the `ToSocketAddrsFuture` by the lifetime in `&self` so that it becomes `ToSocketAddrsFuture<'_>`.

The reason is that this method will become an `async fn` at some point and the returned future in `async fn` gets desugared so that it borrows all function arguments (or, rather, their lifetimes).

Reply:
dirty code in this commit: https://github.com/async-rs/async-std/pull/74/commits/83636b9bfed28a4bbadd0f271cdfc14a520612fd

I'm not sure is that a code style?